src/internal_modules/roc_fec/reader.cpp

```diff
@@ -59,26 +60,35 @@ bool Reader::is_alive() const {
     return alive_;
 }
 
-packet::PacketPtr Reader::read() {
+int Reader::read(packet::PacketPtr& pp) {
     roc_panic_if_not(is_valid());
+
     if (!alive_) {
-        return NULL;
+        return error::ErrNoData;
     }
-    packet::PacketPtr pp = read_();
-    if (pp) {
+
+    int code = read_(pp);
+    if (code == 0) {
         n_packets_++;
     }
-    // check if alive_ has changed
-    return (alive_ ? pp : NULL);
+    if (!alive_) {
+        code = error::ErrNoData;
+        pp = NULL;
+    }
```

Comment:
nitpick: Since this is failure branch, it may be cleaner to use early return

```cpp
if (!alive_) {
    pp = NULL;
    return error::ErrNoData;
}
```